### PR TITLE
chore(debugger): add support for multiple remote config configurations

### PR DESCRIPTION
**Pull request by shatzi**
## Description
debugger can handle multiple configurations from remote config by include merge all probes gotten from remote config

## Motivation
remote config can send multiple configurations to tracer. previously the tracer only uses probe from the last configuration received. now, all probes from those configurations are applied

## Design 
the debugger keep a dict for config-id and applied probes from that config to keep trace on each config changes.

## Testing strategy
automated tests verify apply new and configuration, changed or removed configuration.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added and follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines), or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/debugging/_debugger.py [hunk 1]
```diff
@@ -229,7 +230,7 @@ def __init__(self, tracer=None):
         )
 
         # Register the debugger with the RCM client.
-        RemoteConfig.register("LIVE_DEBUGGING", self.__rc_adapter__(self._on_configuration))
+        RemoteConfig.register(DEBUGGER_PRODUCT, self.__rc_adapter__(self._on_configuration))
```
Comment (P403n1x87) [inline ddtrace/debugging/_debugger.py:233]: Perhaps we should make this change in a separate PR

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 2]
```diff
@@ -119,13 +121,32 @@ def probe(_id, _type, attribs):
     raise ValueError("Unknown probe type: %s" % _type)
 
 
+_SNAPSHOT_PREFIX = "snapshotProbe_"
+_METRIC_PREFIX = "metricProbe_"
+_CONFIG_REGEX = re.compile(r"^[\da-f]{8}-([\da-f]{4}-){3}[\da-f]{12}$", re.IGNORECASE)
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:126]: Can we leave an example of a config ID as a comment here?

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 3]
```diff
@@ -119,13 +121,32 @@ def probe(_id, _type, attribs):
...
+
+
+def _makeProbes(probes, _type):
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:129]:
```suggestion
def _make_probes(probes, _type):
```

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 4]
```diff
@@ -119,13 +121,32 @@ def probe(_id, _type, attribs):
...
+def get_probes(config_id, config):
+    # type: (str, dict) -> Iterable[Probe]
+
+    if config_id.startswith(_SNAPSHOT_PREFIX):
+        return _makeProbes([config], "snapshotProbes")
+
+    if config_id.startswith(_METRIC_PREFIX):
+        return chain(_makeProbes([config], "metricProbes"))
+
+    if _CONFIG_REGEX.match(config_id):
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:143]: How big of a deal is if we don't check for the expected pattern here? I'd like to avoid the call to `match` if we can.

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 5]
```diff
@@ -158,34 +179,17 @@ class ProbeRCAdapter(object):
...
-        probes = get_probes(config)
-
-        current_probes = {_.probe_id: _ for _ in probes}
-
-        new_probes = [p for _, p in current_probes.items() if _ not in self._probes]
-        deleted_probes = [p for _, p in self._probes.items() if _ not in current_probes]
-        modified_probes = [
-            p for _, p in current_probes.items() if _ in self._probes and probe_modified(p, self._probes[_])
-        ]
+    def _update_probes(self, prev_probes, next_probes):
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:189]:
```suggestion
    def _dispatch_prove_events(self, prev_probes, next_probes):
```
WDYT?

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 6]
```diff
@@ -194,4 +198,33 @@ def __call__(self, metadata, config):
...
+            log.debug("Emitting probe status log messages")
+            probes = [probe for config in self._configs.values() for probe in config.values()]
+            self._callback(ProbePollerEvent.STATUS_UPDATE, probes)
+            self._next_status_update_timestamp()
+
+        if metadata is None:
+            log.warning("no metadata was provided")
+            return
+        else:
+            self._update_probes_for_config(metadata.id, config)
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:230]:
```suggestion

        self._update_probes_for_config(metadata.id, config)
```

File: tests/debugging/probe/test_remoteconfig.py [hunk 7]
```diff
@@ -177,3 +185,79 @@ def cb(e, ps):
...
+        events.add((e, frozenset([p.probe_id if isinstance(p, Probe) else p for p in ps])))
+
+    def validate_events(expected):
+        assert events == expected
+        events.clear()
+
+    old_interval = config.diagnostics_interval
+    config.diagnostics_interval = 0.5
+    try:
+        addapter = ProbeRCAdapter(cb)
```
Comment (P403n1x87) [inline tests/debugging/probe/test_remoteconfig.py:203]:
```suggestion
        adapter = ProbeRCAdapter(cb)
```

File: tests/debugging/probe/test_remoteconfig.py [hunk 8]
```diff
@@ -177,3 +185,79 @@ def cb(e, ps):
         }
     finally:
         config.diagnostics_interval = old_interval
+
+
+def test_multiple_configs():
+    events = set()
+
+    def cb(e, ps):
+        events.add((e, frozenset([p.probe_id if isinstance(p, Probe) else p for p in ps])))
```
Comment (P403n1x87) [inline tests/debugging/probe/test_remoteconfig.py:194]:
```suggestion
        events.add((e, {p.probe_id if isinstance(p, Probe) else p for p in ps}))
```
there usually is no need for   `frozenset`s in tests.

File: tests/debugging/probe/test_remoteconfig.py [hunk 9]
```diff
@@ -177,3 +185,79 @@ def cb(e, ps):
...
+                "id": "probe1",
+                "active": True,
+                "tags": ["boo:far"],
+                "where": {"sourceFile": "tests/debugger/submod/stuff.py", "lines": ["36"]},
+            },
+        )
+
+        validate_events(
+            {
+                (ProbePollerEvent.NEW_PROBES, frozenset(["probe1"])),
```
Comment (P403n1x87) [inline tests/debugging/probe/test_remoteconfig.py:217]:
```suggestion
                (ProbePollerEvent.NEW_PROBES, {"probe1"}),
```

**Review by P403n1x87 — CHANGES_REQUESTED** [review]
🙌  that's great, thanks for working on this! Just a few comments, otherwise LGTM in general.

**Comment by P403n1x87** [comment]
Requires #4631

File: ddtrace/debugging/_probe/remoteconfig.py (hunk 2 above)
Comment (shatzi) [reply ddtrace/debugging/_probe/remoteconfig.py:126]: We will be able to do once the old configuration get a prefix as well. but currently we need to make sure that different prefixes are not been treated as service config.

File: ddtrace/debugging/_probe/remoteconfig.py (hunk 4 above)
Comment (shatzi) [reply ddtrace/debugging/_probe/remoteconfig.py:143]: we can avoid it once we no longer need to support this configuration type. but until then we need to verify we only treat configurations id that are guid as service config.

File: ddtrace/debugging/_debugger.py (hunk 1 above)
Comment (shatzi) [reply ddtrace/debugging/_debugger.py]: sounds good.

File: tests/debugging/probe/test_remoteconfig.py (hunk 7 above)
Comment (shatzi) [reply tests/debugging/probe/test_remoteconfig.py]: changed

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 10]
```diff
@@ -158,40 +179,52 @@ class ProbeRCAdapter(object):
...
         self._callback = callback
-        self._probes = {}  # type: Dict[str, Probe]
+        self._configs = {}  # type: Dict[str, Dict[str, Probe]]
         self._next_status_update_timestamp()
 
     def _next_status_update_timestamp(self):
         # type: () -> None
         self._status_timestamp = time.time() + config.diagnostics_interval
 
+    def _dispatch_prove_events(self, prev_probes, next_probes):
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:189]:
```suggestion
    def _dispatch_probe_events(self, prev_probes, next_probes):
```

File: ddtrace/debugging/_probe/remoteconfig.py [hunk 11]
```diff
@@ -158,40 +179,52 @@ class ProbeRCAdapter(object):
...
+            self._callback(ProbePollerEvent.NEW_PROBES, new_probes)
+
+    def _update_probes_for_config(self, config_id, config):
+        # type: (str, Any) -> None
+        prev_probes = self._configs.get(config_id, {})  # type: Dict[str, Probe]
+        next_probes = (
+            {probe.probe_id: probe for probe in get_probes(config_id, config)} if config is not None else {}
+        )  # type: Dict[str, Probe]
+
+        self._dispatch_prove_events(prev_probes, next_probes)
```
Comment (P403n1x87) [inline ddtrace/debugging/_probe/remoteconfig.py:208]:
```suggestion
        self._dispatch_probe_events(prev_probes, next_probes)
```

**Review by P403n1x87 — CHANGES_REQUESTED** [review]
🙌  awesome, thanks! Just a typo, otherwise LGTM!

**Review by mabdinur — COMMENTED** [review]
does this change need a release note?

**Comment by P403n1x87** [comment]
> does this change need a release note?

This change is required to deal with internal details, so we should be good without one.

**Review by mabdinur — APPROVED** [review]
